src/sources/prometheus/scrape.rs

```diff
@@ -252,33 +231,19 @@ struct EndpointInfo {
     honor_label: bool,
 }
 
-async fn prometheus(
-    config: PrometheusScrapeConfig,
-    urls: Vec<http::Uri>,
-    tls: TlsSettings,
-    proxy: ProxyConfig,
-    shutdown: ShutdownSignal,
-    mut out: SourceSender,
-) -> Result<(), ()> {
-    let mut stream = IntervalStream::new(tokio::time::interval(Duration::from_secs(
-        config.scrape_interval_secs,
-    )))
-    .take_until(shutdown)
-    .map(move |_| stream::iter(urls.clone()))
-    .flatten()
-    .map(move |url| {
-        let client = HttpClient::new(tls.clone(), &proxy).expect("Building HTTP client failed");
-        let endpoint = url.to_string();
-
-        let mut request = Request::get(&url)
-            .header(http::header::ACCEPT, "text/plain")
-            .body(Body::empty())
-            .expect("error creating request");
-        if let Some(auth) = &config.auth {
-            auth.apply(&mut request);
-        }
+#[derive(Clone)]
+struct PrometheusScrapeContext {
+    honor_labels: bool,
+    instance_tag: Option<String>,
+    endpoint_tag: Option<String>,
+    instance_info: Option<InstanceInfo>,
+    endpoint_info: Option<EndpointInfo>,
+}
 
-        let instance_info = config.instance_tag.as_ref().map(|tag| {
+impl HttpScraper for PrometheusScrapeContext {
+    /// Builds the instance info and endpoint info for the current request
+    fn build(&mut self, url: &Uri) {
+        self.instance_info = self.instance_tag.as_ref().map(|tag| {
```

Comment:
This form of mutable state is somewhat undesirable. IMO the ideal would be for `fn build` to return a state object, which would in turn implement `fn on_response` etc. Does that make sense, or is that a lot of work?

Reply:
Makes sense, not a lot of work. I took a stab at it by "consuming" the self. Didn't see a better way.